stdlib/public/core/StringUTF16.swift

```diff
@@ -45,24 +47,28 @@ extension String {
     // TODO: swift-3-indexing-model - add docs
     @warn_unused_result
     public func next(i: Index) -> Index {
+      // FIXME: swift-3-indexing-model: range check i?
       return Index(_offset: i._offset.advanced(by: 1))
```

Comment:
@gribozavr 

I am wondering what you are thinking for bounds checking of an index when advancing it in some way especially when Indexes are Int (or Int like). For example something like the following with a pre and post condition check? What to do about the boundary conditions with overflow and underflow?

```
    @warn_unused_result
    public func next(i: Index) -> Index {
      _rangeCheck(i._offset, bounds: 0..<_length)
      let offset = _unsafePlus(i._offset, 1)
      _rangeCheck(offset, bounds: 0..<_length)
      return Index(_offset: offset)
    }
    public func _rangeCheck(index: int, bounds: Range<Int>) {
      _precondition(bounds.startIndex <= index, "blah")
      _precondition(index < bounds.endIndex, "blah")
    }
```

(ignoring code reuse across advance, previous, etc)